sway/commands/output/background.c

```diff
@@ -117,40 +105,47 @@ struct cmd_results *output_cmd_background(int argc, char **argv) {
 		}
 
 		bool can_access = access(src, F_OK) != -1;
-		if (!can_access) {
-			sway_log_errno(SWAY_ERROR, "Unable to access background file '%s'",
-					src);
-			config_add_swaynag_warning("Unable to access background file '%s'",
-					src);
-			struct cmd_results *result = cmd_results_new(CMD_FAILURE,
-					"unable to access background file '%s'", src);
-			free(src);
-			return result;
-		} else {
-			output->background = src;
-			output->background_option = strdup(mode);
-		}
 		argc -= j + 1; argv += j + 1;
 
-		output->background_fallback = NULL;
+		char *fallback = NULL;
 		if (argc && *argv[0] == '#') {
-			if (!validate_color(argv[0])) {
-				return cmd_results_new(CMD_INVALID,
-						"fallback color should be of the form #RRGGBB");
+			if (validate_color(argv[0])) {
+				fallback = strdup(argv[0]);
+				if (!fallback) goto cleanup;
+			} else {
+				sway_log(SWAY_ERROR, "fallback '%s' should be of the form #RRGGBB", argv[0]);
+				config_add_swaynag_warning("fallback '%s' should be of the form #RRGGBB\n", argv[0]);
 			}
-
-			output->background_fallback = strdup(argv[0]);
 			argc--; argv++;
+		}
 
-			if (!can_access) {
-				output->background = output->background_fallback;
-				output->background_option = strdup("solid_color");
-				output->background_fallback = NULL;
+		free(output->background_option);
+		free(output->background_fallback);
+		if (!can_access) {
+			if (!fallback) {
```

Comment:
To avoid use-after-free I'd recommend setting `output->background_option` and `output->background_fallback` to NULL immediately after freeing them -- otherwise they may be used without being replaced if `can_access` is false and `fallback` is NULL.